pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,907 @@
+	pv.Annotations[pvProvisioningRequired] = pvProvisioningCompleted
+	pv, err = controller.client.UpdatePersistentVolume(pv)
+	if err != nil {
+		// rollback
+		pv.Annotations[pvProvisioningRequired] = oldValue
+		// TODO:  https://github.com/kubernetes/kubernetes/issues/14443
+		// the volume was created in the infrastructure and likely has a PV name on it, but we failed to mark the provisioning completed.
+		return
+	}
+}
+
+// recycleVolume recycles a volume that has been released from its claim.
+// This func (and Provision/Delete) locks on the PV.Name to limit 1 operation per volume at a time.
+func recycleVolume(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+	controller.locks[pv.Name].Lock()
+	defer func(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+		controller.locks[pv.Name].Unlock()
+		delete(controller.locks, pv.Name)
+	}(pv, controller)
+
+	if isAnnotationMatch(pvRecycleRequired, pvRecycleCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is already recycled.", pv.Name)
+		return
+	}
+
+	spec := volume.NewSpecFromPersistentVolume(pv, false)
+	plugin, err := controller.pluginMgr.FindRecyclablePluginBySpec(spec)
+	if err != nil {
+		glog.Errorf("Could not find recyclable volume plugin for spec: %+v", err)
+		return
+	}
+	recycler, err := plugin.NewRecycler(spec)
+	if err != nil {
+		glog.Errorf("could not obtain Recycler for spec: %+v", err)
+		return
+	}
+	// blocks until completion
+	err = recycler.Recycle()
+	if err != nil {
+		glog.Errorf("PersistentVolume[%s] failed recycling: %+v", pv.Name, err)
+		pv.Status.Message = fmt.Sprintf("Recycling error: %s", err)
+		return
+	}
+
+	oldValue := pv.Annotations[pvRecycleRequired]
+	pv.Annotations[pvRecycleRequired] = pvRecycleCompleted
+	_, err = controller.client.UpdatePersistentVolume(pv)
+	if err != nil {
+		glog.Errorf("PersistentVolume[%s] failed update: %#v", err)
+		// rollback on the pv pointer
+		pv.Annotations[pvRecycleRequired] = oldValue
+		return
+	}
+	glog.V(5).Infof("PersistentVolume[%s] successfully recycled through plugin\n", pv.Name)
+
+}
+
+// deleteVolume removes a volume from the storage provider.  The volume has been released from its claim.
+// This func (and Provision/Recycle) locks on the PV.Name to limit 1 operation per volume at a time.
+func deleteVolume(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+	controller.locks[pv.Name].Lock()
+	defer func(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+		controller.locks[pv.Name].Unlock()
+		delete(controller.locks, pv.Name)
+	}(pv, controller)
+
+	if isAnnotationMatch(pvDeleteRequired, pvDeleteCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is already deleted", pv.Name)
+	}
+
+	spec := volume.NewSpecFromPersistentVolume(pv, false)
+	plugin, err := controller.pluginMgr.FindDeletablePluginBySpec(spec)
+	if err != nil {
+		glog.Errorf("Could not find deletable volume plugin for spec: %+v", err)
```

Comment:
Like in `provisionVolume()`, all errors in this function (not only this one) should propagate to pv so the user knows it won't ever succeed.